lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package storage
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/ChainSafe/gossamer/pkg/trie"
+)
+
+// storageDiff is a structure that stores the differences between consecutive
+// states of a trie, such as those occurring during the execution of a block.
+// It records updates (upserts), deletions, and changes to child tries.
+// This mechanism facilitates applying state transitions efficiently.
+// Changes accumulated in storageDiff can be applied to a trie using
+// the `applyToTrie` method
+type storageDiff struct {
+	upserts        map[string][]byte
+	deletes        map[string]bool
+	childChangeSet map[string]*storageDiff
+	mtx            sync.RWMutex
```

Comment:
is this structure used among different goroutines? from what I understand this structure should be used only when a block is being executed which means that this should be only accessible in the block context and once the execution finishes it is no longer needed, right?

Reply:
I added this mtx to do it "thread safe" but currently this is only used by `TrieState` and `TrieState` already has a mutex, I believe we can remove the mtx from `StorageDiff` if you don't see any issue with that

Reply:
Yeah, if we can lock on trie state than we can avoid locking here

Reply:
Done! thks